polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -47,38 +47,112 @@ namespace polygeist {
 
 using namespace mlir;
 
+namespace {
+
 //===----------------------------------------------------------------------===//
 // Helper Functions
 //===----------------------------------------------------------------------===//
 
-namespace {
+/// Returns true if the callable operation \p callableOp has a callable region
+/// and private visibility, and false otherwise.
+bool isPrivateCallable(CallableOpInterface callableOp) {
+  auto sym = dyn_cast<SymbolOpInterface>(callableOp.getOperation());
+  return (sym && sym.isPrivate() && callableOp.getCallableRegion());
+}
 
-/// Represent a candidate for the argument promotion transformation.
-class Candidate {
+/// Returns true if the call \p callOp is recursive, and false otherwise.
+bool isRecursiveCall(CallOpInterface callOp) {
+  auto callable = dyn_cast<CallableOpInterface>(callOp.resolveCallable());
+  return (callable && callable.getCallableRegion() &&
+          callable.getCallableRegion()->isAncestor(callOp->getParentRegion()));
+}
+
+/// Returns true if \p type is 'memref<?xstruct<>>', and false otherwise.
+bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  auto structType = dyn_cast<LLVM::LLVMStructType>(mt.getElementType());
+  if (!structType || llvm::any_of(structType.getBody(), [](Type memType) {
+        return isa<LLVM::LLVMStructType>(memType);
+      }))
+    return false;
+  return true;
+}
+
+// Returns true if the block of \p startOp contains an instruction (after \p
+// startOp) that has a side effects on \p val.
+auto existsSideEffectAfter(Value val, Operation *startOp) {
+  assert(startOp && "Expecting a valid pointer");
+  Block *block = startOp->getBlock();
+
+  for (Operation &op : *block) {
+    if (op.isBeforeInBlock(startOp) || isMemoryEffectFree(&op))
+      continue;
+
+    // Operations with unknown side effects might write to any memory.
```

Comment:
according to the comment, it may write to any memory, then why `continue` instead of `return true`?